File: mobile/test/ayirici_tutarlilik_test.dart
// (P77) IKI AYRISTIRICI, TEK AYIRICI KURALI.
//
// `core/para.dart` (kurus tam sayisi) ve `core/sayi.dart` (ondalikli
// deger) AYRI donus tipleri uretir ama AYIRICI KURALI ayni olmak
// zorundadir: kullanici metrekare alaninda ve tutar alaninda ayni yazimi
// kullanabilmeli. Ikisi ayrisirsa, ayni sitede ayni metin iki farkli sayi
// girer — P49/P50'nin panel-mobil arasinda buldugu kusurun ayni evin
// icindeki hali.
//
// Bu test DEGERLERI degil, KABUL/RED kararini karsilastirir: para
// ayristiricisi iki basamak kisiti ve isaret POLITIKASI tasir (negatif
// reddedilir), sayi ayristiricisi tasimaz. Ortusen alanda ikisi AYNI
// karari vermelidir.
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/src/core/para.dart';
import 'package:mobile/src/core/sayi.dart';

void main() {
  // Politika farki OLMAYAN girdiler: pozitif, en fazla iki ondalik hane.
  const kabulEdilmeli = [
    '1250', '1250,50', '1250.50', '1.250', '1.250,00', '0', '0,05',
  ];
  const reddedilmeli = [
    '', '  ', 'abc', '1250,', ',50', '1250.', '.50', '1 2 3',
    // `1,234` BURADA DEGIL: para onu reddeder (kurus iki hanedir) ama
    // sayi kabul eder (1,234). Bu bir AYIRICI farki degil, POLITIKA
    // farkidir ve asagida ayrica test edilir. Ortak listeye koymak,
    // dogru davranisi tutarsizlik gibi gostermek olurdu.
  ];

  group('ayirici kurali iki ayristiricida AYNI', () {
    for (final g in kabulEdilmeli) {
      test('KABUL: "$g"', () {
        expect(tlMetniniKurusaCevir(g), isNotNull, reason: 'para');
        expect(sayiCoz(g).tur, SayiTuru.sayi, reason: 'sayi');
      });
    }

    for (final g in reddedilmeli) {
      test('RED: "$g"', () {
        expect(tlMetniniKurusaCevir(g), isNull, reason: 'para');
        // Bos girdi `sayiCoz`ta AYRI bir durumdur (bos != gecersiz);
        // ortak olan sey "sayi URETMEZ"dir.
        expect(sayiCoz(g).tur, isNot(SayiTuru.sayi), reason: 'sayi');
      });
    }
  });

  group('politika farki BILINCLI', () {
    test('negatif: para REDDEDER, sayi kabul eder', () {
      // Isaret bir BICIM degil ALAN kuralidir: tutar negatif olamaz ama
      // bir olcu/fark negatif olabilir. Bu ayrim kasitlidir.
      expect(tlMetniniKurusaCevir('-5'), isNull);
      expect(sayiCoz('-5').deger, -5);
    });

    test('uc ondalik hane: para REDDEDER (kurus iki hanedir)', () {
      expect(tlMetniniKurusaCevir('1,234'), isNull);
    });

    test('bos girdi: sayi BOS der, para null', () {
      expect(sayiCoz('').tur, SayiTuru.bos);
      expect(tlMetniniKurusaCevir(''), isNull);
    });
  });
}

File: admin-web/tests/ayirici-tutarlilik.test.ts
// (P78) IKI AYRISTIRICI, TEK AYIRICI KURALI — panelin karsiligi.
//
// P77 bunu mobilde kilitledi. Panelde de ayni ikilik var: tutar alanlari
// `lib/money.ts::tlToKurus`, para OLMAYAN sayilar `lib/sayi.ts::sayiCoz`
// kullaniyor. Ikisinin KENDI testleri vardi; ARALARINDAKI sozlesmenin
// testi yoktu — biri "sadelestirilirse" digeriyle sessizce ayrisir ve
// kullanici ayni yazimi iki alanda kullanamaz olurdu.
//
// Kilit KABUL/RED kararini karsilastirir, degeri degil: donus tipleri
// farkli (kurus tamsayisi ↔ ondalikli sayi).
import { describe, expect, it } from "vitest";

import { tlToKurus } from "@/lib/money";
import { sayiCoz } from "@/lib/sayi";

// Politika farki OLMAYAN girdiler: pozitif, en fazla iki ondalik hane.
const KABUL = ["1250", "1250,50", "1250.50", "1.250", "1.250,00", "0", "0,05"];
const RED = ["", "   ", "abc", "1250,", ",50", "1250.", ".50", "1 2 3"];

describe("ayirici kurali iki ayristiricida AYNI", () => {
  it.each(KABUL)("KABUL: %s", (girdi) => {
    expect(tlToKurus(girdi), "para").not.toBeNull();
    expect(sayiCoz(girdi).tur, "sayi").toBe("sayi");
  });

  it.each(RED)("RED: %s", (girdi) => {
    expect(tlToKurus(girdi), "para").toBeNull();
    // Bos girdi `sayiCoz`ta AYRI bir durumdur (bos != gecersiz); ortak
    // olan sey "sayi URETMEZ"dir.
    expect(sayiCoz(girdi).tur, "sayi").not.toBe("sayi");
  });
});

describe("politika farklari BILINCLI", () => {
  it("negatif: para REDDEDER, sayi kabul eder", () => {
    // Isaret bir BICIM degil ALAN kuralidir: tutar negatif olamaz ama bir
    // olcu/fark olabilir (mobilde ayni karar — P77).
    expect(tlToKurus("-5")).toBeNull();
    expect(sayiCoz("-5")).toEqual({ tur: "sayi", deger: -5 });
  });

  it("uc ondalik hane: para REDDEDER (kurus iki hanedir)", () => {
    expect(tlToKurus("1,234")).toBeNull();
    expect(sayiCoz("1,234")).toEqual({ tur: "sayi", deger: 1.234 });
  });

  it("bos girdi: sayi BOS der, para null", () => {
    expect(sayiCoz("")).toEqual({ tur: "bos" });
    expect(tlToKurus("")).toBeNull();
  });
});

describe("panel ve mobil AYNI kurali uygular", () => {
  // `mobile/test/ayirici_tutarlilik_test.dart` ayni listeleri surer.
  // Listeler ayrisirsa bu yorum yalan olur; bu yuzden liste degerleri
  // ORADA DA aynen duruyor ve degistirilirken ikisi birlikte
  // degistirilmelidir.
  it("kabul/red listeleri mobil testiyle AYNI degerleri tasir", () => {
    expect(KABUL).toEqual([
      "1250", "1250,50", "1250.50", "1.250", "1.250,00", "0", "0,05",
    ]);
    expect(RED).toEqual([
      "", "   ", "abc", "1250,", ",50", "1250.", ".50", "1 2 3",
    ]);
  });
});
